app/controllers/application_controller.rb

```diff
@@ -21,10 +21,10 @@ def self.allow_unauthenticated!(*actions)
       define_method action do
         if user_logged_in?
           send("authenticated_#{action}")
-          render action: "#{action}/authenticated"
+          render action: "#{action}/authenticated" unless performed?
```

Comment:
This allows us to override the render call in the individual controllers and only do **this** render if we've not rendered elsewhere.

Reply:
I'm a bit confused. Won't overriding this function in the controller not run this code?

Something like:
```
class Tracks::ConceptsController < ApplicationController
  before_action :use_track

  allow_unauthenticated! :index

  def external_index
    // This code runs, and not the method generated by the macro
  end
end
```

Reply:
The method generated by the macro is `index`.

So you get:
```ruby
def index
  if user_logged_in?
    send("authenticated_index")
    render action: "index/authenticated" unless performed?
  else
    send("external_index")
    render action: "index/external" unless performed?
  end
end```

So this generated method calls `authenticated_index` (setting instance variables etc) then renders `index/authenticated`, unless the `authenticated_index` method itself calls render or redirect.

Reply:
Ah!